File: packages/typescript/src/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.ts
import type { AlvoCriptografico } from '../AlvoCriptografico.ts';
import type { AtaqueInterface } from '../AtaqueInterface.ts';
import { ResultadoAtaque } from '../ResultadoAtaque.ts';
import { SkipAtaqueException } from '../SkipAtaqueException.ts';
import { randomInt } from '../Util.ts';

/**
 * Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
 * o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
 * posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
 */
export class AtaqueCoberturaDependenciaIV implements AtaqueInterface {
  private readonly tamanhoBloco: number;
  private readonly perturbacoesPorPosicao: number;

  constructor({ tamanhoBloco = 32, perturbacoesPorPosicao = 5 }: { tamanhoBloco?: number; perturbacoesPorPosicao?: number } = {}) {
    this.tamanhoBloco = tamanhoBloco;
    this.perturbacoesPorPosicao = perturbacoesPorPosicao;
  }

  nome(): string {
    return 'Cobertura de dependência do IV (entrada x saída)';
  }

  executar(alvo: AlvoCriptografico): ResultadoAtaque {
    const chave = alvo.chaveDeTeste();
    const tamanhoIv = alvo.tamanhoIv();
    const ivBase = Buffer.alloc(tamanhoIv);

    const ksBase: Buffer | null = alvo.gerarKeystreamBruto(chave, ivBase, 'enc', this.tamanhoBloco);
    if (ksBase == null) {
      throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
    }

    const paresIndependentes: string[] = [];

    for (let posIv = 0; posIv < tamanhoIv; posIv++) {
      const afetou = new Array<boolean>(this.tamanhoBloco).fill(false);

      for (let p = 0; p < this.perturbacoesPorPosicao; p++) {
        const iv = Buffer.from(ivBase);
        iv[posIv] = randomInt(1, 255);
        const ks = alvo.gerarKeystreamBruto(chave, iv, 'enc', this.tamanhoBloco);

        for (let posSaida = 0; posSaida < this.tamanhoBloco; posSaida++) {
          if (ks[posSaida] !== ksBase[posSaida]) {
            afetou[posSaida] = true;
          }
        }
      }

      afetou.forEach((ok, posSaida) => {
        if (!ok) {
          paresIndependentes.push(`iv[${posIv}] -> saida[${posSaida}]`);
        }
      });
    }

    if (paresIndependentes.length > 0) {
      return new ResultadoAtaque(
        this.nome(),
        true,
        'alta',
        `${paresIndependentes.length} par(es) sem dependência detectável em ${this.perturbacoesPorPosicao} tentativas cada`,
        { pares: paresIndependentes.slice(0, 20) },
      );
    }

    return new ResultadoAtaque(
      this.nome(),
      false,
      'info',
      `Todas as ${tamanhoIv} posições do IV influenciam todas as ${this.tamanhoBloco} posições de saída`,
    );
  }
}

File: packages/dart/lib/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
/// o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
/// posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
class AtaqueCoberturaDependenciaIV extends Ataque {
  final int _tamanhoBloco;
  final int _perturbacoesPorPosicao;

  AtaqueCoberturaDependenciaIV({
    this._tamanhoBloco = 32,
    this._perturbacoesPorPosicao = 5,
  });

  @override
  String nome() => 'Cobertura de dependência do IV (entrada x saída)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = alvo.chaveDeTeste();
    final tamanhoIv = alvo.tamanhoIv();
    final ivBase = Uint8List(tamanhoIv);

    final ksBase = alvo.gerarKeystreamBruto(
      chave,
      ivBase,
      'enc',
      _tamanhoBloco,
    );

    final paresIndependentes = <String>[];

    for (var posIv = 0; posIv < tamanhoIv; posIv++) {
      final afetou = List<bool>.filled(_tamanhoBloco, false);

      for (var p = 0; p < _perturbacoesPorPosicao; p++) {
        final iv = Uint8List.fromList(ivBase);
        iv[posIv] = randomInt(1, 255);
        final ks = alvo.gerarKeystreamBruto(
          chave,
          iv,
          'enc',
          _tamanhoBloco,
        );

        for (var posSaida = 0; posSaida < _tamanhoBloco; posSaida++) {
          if (ks[posSaida] != ksBase[posSaida]) {
            afetou[posSaida] = true;
          }
        }
      }

      for (var posSaida = 0; posSaida < _tamanhoBloco; posSaida++) {
        if (!afetou[posSaida]) {
          paresIndependentes.add('iv[$posIv] -> saida[$posSaida]');
        }
      }
    }

    if (paresIndependentes.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.alta,
        '${paresIndependentes.length} par(es) sem dependência detectável em '
            '$_perturbacoesPorPosicao tentativas cada',
        {'pares': paresIndependentes.take(20).toList()},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Todas as $tamanhoIv posições do IV influenciam todas as '
          '$_tamanhoBloco posições de saída',
    );
  }
}
